In [1]:
from pathlib import Path
import pandas as pd

ROOT = Path.cwd().parent   # notebook berjalan dari folder notebooks/, jadi naik satu level ke root
DATA_DIR = ROOT / "data/pasar_tradisional/Cleaned_After_Imputation"

files = sorted(DATA_DIR.glob("*.csv"))
print(len(files), "file ditemukan")
print(files[0].name)

10 file ditemukan
final_komoditas_tradisional_bawang_merah_2022_2026.csv


In [2]:
df = pd.read_csv(files[0])
print(df.shape)
df.dtypes

(1415, 35)


Date_Param                       str
Aceh                         float64
Bali                         float64
Banten                       float64
Bengkulu                     float64
DI Yogyakarta                float64
DKI Jakarta                  float64
Gorontalo                    float64
Jambi                        float64
Jawa Barat                   float64
Jawa Tengah                  float64
Jawa Timur                   float64
Kalimantan Barat             float64
Kalimantan Selatan           float64
Kalimantan Tengah            float64
Kalimantan Timur             float64
Kalimantan Utara             float64
Kepulauan Bangka Belitung    float64
Kepulauan Riau               float64
Lampung                      float64
Maluku                       float64
Maluku Utara                 float64
Nusa Tenggara Barat          float64
Nusa Tenggara Timur          float64
Papua                        float64
Papua Barat                  float64
Riau                         float64
S

In [3]:
display(df.head(10))
df.isna().sum()

,Date_Param,Aceh,Bali,Banten,Bengkulu,DI Yogyakarta,DKI Jakarta,Gorontalo,Jambi,Jawa Barat,...,Papua Barat,Riau,Sulawesi Barat,Sulawesi Selatan,Sulawesi Tengah,Sulawesi Tenggara,Sulawesi Utara,Sumatera Barat,Sumatera Selatan,Sumatera Utara
0,2022-01-01,32000.0,22000.0,27400.0,30000.0,26000.0,34150.0,35850.0,24500.0,27350.0,...,46250.0,28400.0,29250.0,25150.0,30750.0,32150.0,33400.0,26250.0,31000.0,30150.0
1,2022-01-02,32000.0,22000.0,27400.0,30000.0,26000.0,34150.0,35850.0,24500.0,27350.0,...,46250.0,28400.0,29250.0,25150.0,30750.0,32150.0,33400.0,26250.0,31000.0,30150.0
2,2022-01-03,32100.0,21750.0,27250.0,29750.0,26250.0,32500.0,38150.0,24500.0,27300.0,...,46250.0,28450.0,24350.0,25200.0,33250.0,32150.0,33400.0,26250.0,31250.0,30500.0
3,2022-01-04,31650.0,21750.0,27250.0,29500.0,26250.0,32500.0,38150.0,24500.0,27100.0,...,46250.0,28450.0,24350.0,25350.0,33250.0,32150.0,34250.0,26250.0,30250.0,30150.0
4,2022-01-05,31000.0,22250.0,27650.0,29750.0,26250.0,32500.0,40650.0,24500.0,27200.0,...,47500.0,28450.0,24350.0,25300.0,33250.0,32150.0,34250.0,26250.0,30750.0,30500.0
5,2022-01-06,31000.0,22000.0,27250.0,29500.0,26250.0,32500.0,41850.0,24500.0,27200.0,...,47500.0,28200.0,24350.0,25300.0,33250.0,32150.0,34650.0,26250.0,30000.0,30250.0
6,2022-01-07,31100.0,22250.0,27250.0,29500.0,26250.0,32500.0,42500.0,24500.0,27000.0,...,47500.0,28200.0,24650.0,25150.0,33250.0,32150.0,35650.0,26250.0,30000.0,30150.0
7,2022-01-08,31100.0,22250.0,27250.0,29500.0,26250.0,32500.0,42500.0,24500.0,27000.0,...,47500.0,28200.0,24650.0,25150.0,33250.0,32150.0,35650.0,26250.0,30000.0,30150.0
8,2022-01-09,31100.0,22250.0,27250.0,29500.0,26250.0,32500.0,42500.0,24500.0,27000.0,...,47500.0,28200.0,24650.0,25150.0,33250.0,32150.0,35650.0,26250.0,30000.0,30150.0
9,2022-01-10,31150.0,21750.0,26350.0,29750.0,26250.0,31250.0,43350.0,24500.0,26650.0,...,50000.0,27900.0,24600.0,25650.0,33250.0,32150.0,41150.0,26250.0,30250.0,30150.0


Date_Param                   0
Aceh                         0
Bali                         0
Banten                       0
Bengkulu                     0
DI Yogyakarta                0
DKI Jakarta                  0
Gorontalo                    0
Jambi                        0
Jawa Barat                   0
Jawa Tengah                  0
Jawa Timur                   0
Kalimantan Barat             0
Kalimantan Selatan           0
Kalimantan Tengah            0
Kalimantan Timur             0
Kalimantan Utara             0
Kepulauan Bangka Belitung    0
Kepulauan Riau               0
Lampung                      0
Maluku                       0
Maluku Utara                 0
Nusa Tenggara Barat          0
Nusa Tenggara Timur          0
Papua                        0
Papua Barat                  0
Riau                         0
Sulawesi Barat               0
Sulawesi Selatan             0
Sulawesi Tengah              0
Sulawesi Tenggara            0
Sulawesi Utara               0
Sumatera

In [4]:
print(df.shape)
print(df["Date_Param"].head(5).tolist())
print(df["Date_Param"].tail(5).tolist())

(1415, 35)
['2022-01-01', '2022-01-02', '2022-01-03', '2022-01-04', '2022-01-05']
['2026-02-08', '2026-02-09', '2026-02-10', '2026-02-11', '2026-02-12']


In [5]:
import re

frames = []
for f in files:
    # "final_komoditas_tradisional_bawang_merah_2022_2026.csv" → "bawang_merah"
    komoditas = re.sub(r"^final_komoditas_tradisional_|_\d{4}_\d{4}$", "", f.stem)

    wide = pd.read_csv(f)
    long = wide.melt(id_vars="Date_Param", var_name="provinsi", value_name="harga")
    long["komoditas"] = komoditas
    frames.append(long)

harga = pd.concat(frames, ignore_index=True)
harga = harga.rename(columns={"Date_Param": "tanggal"})
harga["tanggal"] = pd.to_datetime(harga["tanggal"])   # sesuaikan format kalau Cell 4 menunjukkan dd/mm/yyyy

print(harga.shape)
print(harga["komoditas"].unique())
harga.head()

(481236, 4)
<StringArray>
[ 'bawang_merah',  'bawang_putih',         'beras',   'cabai_merah',
   'cabai_rawit',   'daging_ayam',   'daging_sapi',    'gula_pasir',
 'minyak_goreng',    'telur_ayam']
Length: 10, dtype: str


,tanggal,provinsi,harga,komoditas
0,2022-01-01,Aceh,32000.0,bawang_merah
1,2022-01-02,Aceh,32000.0,bawang_merah
2,2022-01-03,Aceh,32100.0,bawang_merah
3,2022-01-04,Aceh,31650.0,bawang_merah
4,2022-01-05,Aceh,31000.0,bawang_merah


In [6]:
n_hari = harga["tanggal"].nunique()
print("hari:", n_hari, "| rentang:", harga["tanggal"].min().date(), "→", harga["tanggal"].max().date())
print("baris aktual:", len(harga), "| ekspektasi:", n_hari * 34 * 10)
print("duplikat grain:", harga.duplicated(["tanggal", "provinsi", "komoditas"]).sum())
print("harga <= 0:", (harga["harga"] <= 0).sum())

hari: 1417 | rentang: 2022-01-01 → 2026-02-12
baris aktual: 481236 | ekspektasi: 481780
duplikat grain: 0
harga <= 0: 0


In [7]:
# tanggal kalender yang tidak ada sama sekali
semua_hari = pd.date_range(harga["tanggal"].min(), harga["tanggal"].max())
hilang = semua_hari.difference(harga["tanggal"].unique())
print(len(hilang), "tanggal hilang")
print(pd.Series(hilang).dt.strftime("%Y-%m-%d (%a)").tolist())

# 16 kombinasi komoditas-tanggal yang hilang
per_kom = harga.groupby("komoditas")["tanggal"].nunique()
print(per_kom[per_kom < harga["tanggal"].nunique()])

87 tanggal hilang
['2022-01-15 (Sat)', '2022-01-16 (Sun)', '2022-01-23 (Sun)', '2022-02-19 (Sat)', '2022-02-20 (Sun)', '2022-03-03 (Thu)', '2022-03-19 (Sat)', '2022-03-20 (Sun)', '2022-04-16 (Sat)', '2022-04-17 (Sun)', '2022-04-23 (Sat)', '2022-04-24 (Sun)', '2022-05-02 (Mon)', '2022-05-03 (Tue)', '2022-05-08 (Sun)', '2022-05-15 (Sun)', '2022-05-26 (Thu)', '2022-06-01 (Wed)', '2022-06-19 (Sun)', '2022-06-25 (Sat)', '2022-06-26 (Sun)', '2022-07-16 (Sat)', '2022-07-17 (Sun)', '2022-07-23 (Sat)', '2022-07-24 (Sun)', '2022-08-20 (Sat)', '2022-08-21 (Sun)', '2022-09-17 (Sat)', '2022-09-18 (Sun)', '2022-09-24 (Sat)', '2022-09-25 (Sun)', '2022-10-16 (Sun)', '2022-10-23 (Sun)', '2022-12-24 (Sat)', '2022-12-25 (Sun)', '2023-02-18 (Sat)', '2023-02-19 (Sun)', '2023-02-25 (Sat)', '2023-02-26 (Sun)', '2023-03-19 (Sun)', '2023-03-25 (Sat)', '2023-03-26 (Sun)', '2023-04-16 (Sun)', '2023-05-20 (Sat)', '2023-05-21 (Sun)', '2023-06-18 (Sun)', '2023-06-24 (Sat)', '2023-06-25 (Sun)', '2023-07-15 (Sat)', '

## Data Quality Findings
- Source data was pre-imputed (missing cells filled); imputation method unknown → volatility may be understated.
- 87 of 1,504 calendar days (5.8%) absent across all commodities — mostly sporadic weekends,
  plus 2022 national holidays (Nyepi, Idulfitri, Ascension Day, Pancasila Day).
- Cabai merah & cabai rawit have 2 extra dates (1,417 vs 1,415 days) not present for other commodities.
- No duplicates at grain (date × province × commodity); no zero or negative prices.
- Implication: day-over-day changes computed via LAG may span multiple days → trends/volatility analyzed at weekly/monthly level.